File: perl/recursed/rcollatz.pl
#! /usr/bin/perl

use warnings;
use strict;

print "Enter your input:\n";
my $input = <STDIN>;
chomp $input;

my @numbers;
my @steps;

for(my $i = 0; $i < 10; $i = $i + 1)
{
    $numbers[$i] = 0;
    $steps[$i] = 0;    
}

my $control = $input;

sub stepCount
{
    my $n = shift;
    if($n == 1)
    {
        return 0;
    }
    else
    {
        if($n % 2 == 0)
        {
            return 1 + stepCount($n/2);
        }
        else
        {
            return 1 + stepCount($n*3+1);
        }
    }   
}
while($control != 1)
{
    $input = $control;
    my $count = stepCount($input);
    my $dup = 0;
    my $dupIndex = 0;
    my $smallIndex = 0;
    my $small = $steps[0];
    for(my $i = 0; $i < 10; $i = $i + 1)
    {
        if($steps[$i] <= $small)
        {
            $small = $steps[$i];
            $smallIndex = $i;
        }
        if($steps[$i] == $count)
        {
            $dup = 1;
            $dupIndex = $i;
        }
    }
    if($count > $steps[$smallIndex] && $dup == 0)    
    {
        $steps[$smallIndex] = $count;
        $numbers[$smallIndex] = $input;
    }
    if($input < $numbers[$dupIndex] && $dup == 1)
    {
        $numbers[$dupIndex] = $input;
    }
    $control = $control - 1;
}

print "Sorted by step count size:\n";
my @index = sort { $steps[$b] <=> $steps[$a] } 0 .. $#steps;

@steps = @steps[@index];
@numbers = @numbers[@index];

for(my $a = 0; $a < 10; $a = $a + 1)
{
    print $numbers[$a], "    ",$steps[$a], "\n";
}

print "Sorted by sequence number size:\n";
my @index2 = sort { $numbers[$b] <=> $numbers[$a] } 0 .. $#numbers;

@steps = @steps[@index2];
@numbers = @numbers[@index2];

for($a = 0; $a < 10; $a = $a + 1)
{
    print $numbers[$a], "    ",$steps[$a], "\n";
}

File: julia/recursed/rcollatz.jl
function stepCount(i::UInt128)
    i == 1 && return 0 
    return i = i&1 == 0 ? 1 + stepCount(i >> 1) : 1 + stepCount(3*i + 1) 
end


sArray = Array{UInt128}(10)
nArray = Array{UInt128}(10)

i = UInt128(10000)
j = UInt128(10000)

while j > 1
    i = j
    ic = UInt128(0)
    ic = stepCount(i)
    dupF = Int8(0)
    dIndex = Int8(1)
    sIndex = Int8(1)
    small = UInt128(sArray[1])
    for k in 1:10
        if(sArray[k] < small)
            small = sArray[k]
            sIndex = k
        end
        if(sArray[k] == ic)
            dupF = 1
            dIndex = k
        end
    end
    if(ic > sArray[sIndex] && dupF == 0)
        sArray[sIndex] = ic
        nArray[sIndex] = i
    end
    if(i < nArray[dIndex] && dupF == 1)
        nArray[dIndex] = i
    end
    j -= 1
end

println("sorted by step count")
 for i in 2:length(sArray)
    for j in 1:length(sArray)-1
      if sArray[j] < sArray[j+1]
          tmp = sArray[j]
          sArray[j] = sArray[j+1]
          sArray[j+1] = tmp
          tep2 = nArray[j]
          nArray[j] = nArray[j+1]
          nArray[j+1] = tep2
      end
    end
  end
for x in 1:10
    println(nArray[x],"     ",sArray[x])
end
println("sorted by number")
for i in 2:length(nArray)
    for j in 1:length(nArray)-1
      if nArray[j] < nArray[j+1]
          tmp = nArray[j]
          nArray[j] = nArray[j+1]
          nArray[j+1] = tmp
          tmp2 = sArray[j]
          sArray[j] = sArray[j+1]
          sArray[j+1] = tmp2
      end
    end
  end


for x in 1:10
    println(nArray[x],"     ",sArray[x])
end
exit(0)


